### ISSUE TITLE
When creating resources in a loop, kube sometimes returns 409's.

### ISSUE THREAD

**Issue by cpressland** [kr8s-org/kr8s#183]
### Which project are you reporting a bug for?

kr8s

### What happened?

I think there's some kind of out of order issue happening but I'm not really sure where to start to track something like this down.

I'm deploying about 20 applications in a for loop, a simplified example more or less looks like this:

```python
from starbug.models.kubernetes.infrastructure.namespace import Namespace
from starbug.models.kubernetes.infrastructure.postgres import Postgres
from starbug.models.kubernetes.infrastructure.rabbitmq import RabbitMQ
from starbug.models.kubernetes.infrastructure.redis import Redis

def main() -> None:
    word1 = choice(["walking", "running", "jumping", "skipping", "hopping"])
    word2 = choice(["red", "blue", "green", "yellow", "orange", "purple", "pink"])
    word3 = choice(["cat", "dog", "bird", "fish", "rabbit", "hamster", "mouse"])

    modules = []

    namespace_name = f"ait-{word1}-{word2}-{word3}"
    modules.append(Namespace(name=namespace_name).complete())
    modules.append(Postgres(namespace=namespace_name).complete())
    modules.append(RabbitMQ(namespace=namespace_name).complete())
    modules.append(Redis(namespace=namespace_name).complete())

    for module in modules:
        for component in module:
            logger.info("Deploying: {}, {}", component.name, component.kind)
            component.create()
```

Each time the `main()` function gets called a new namespace gets created and then the subcomponents of each components get installed one at a time. Here is how we define an application, Redis for example:

```python
"""Define a Redis Instance."""
from kr8s.objects import Deployment, Service, ServiceAccount


class Redis:
    """Define a Redis Instance."""

    def __init__(self, namespace: str, image: str | None = None) -> None:
        """Initialize the Redis class."""
        self.namespace = namespace
        self.image = image or "docker.io/redis:6"
        self.name = "redis"
        self.labels = {"app": "redis"}
        self.serviceaccount = ServiceAccount({
            "apiVersion": "v1",
            "kind": "ServiceAccount",
            "metadata": {
                "name": self.name,
                "namespace": self.namespace,
            },
        })
        self.service = Service({
            "apiVersion": "v1",
            "kind": "Service",
            "metadata": {
                "name": self.name,
                "namespace": self.namespace,
                "labels": self.labels,
            },
            "spec": {
                "ports": [{"port": 6379, "targetPort": 6379}],
                "selector": self.labels,
            },
        })
        self.deployment = Deployment({
            "apiVersion": "apps/v1",
            "kind": "Deployment",
            "metadata": {
                "name": self.name,
                "namespace": self.namespace,
                "labels": self.labels,
            },
            "spec": {
                "replicas": 1,
                "selector": {
                    "matchLabels": self.labels,
                },
                "template": {
                    "metadata": {
                        "labels": self.labels,
                        "annotations": {
                            "kubectl.kubernetes.io/default-container": self.name,
                        },
                    },
                    "spec": {
                        "serviceAccountName": self.name,
                        "containers": [
                            {
                                "name": self.name,
                                "image": self.image,
                                "ports": [{"containerPort": 6379}],
                            },
                        ],
                    },
                },
            },
        })

    def complete(self) -> tuple[ServiceAccount, Service, Deployment]:
        """Return all deployable objects as a tuple."""
        return (self.serviceaccount, self.service, self.deployment)

```

This works really well, _most of the time_.  Unfortunately sometimes Kubernetes seems to return a 409 stating that an object already exists. Well, that's impossible as only a few milliseconds ago did we create the namespace that these components are going to be installed in. I modified the `main()` function with some janky log messages and a pause to try and determine why this is happening.

```python
    for module in modules:
        for component in module:
            try:
                logger.info("Deploying: {}, {}", component.name, component.kind)
                component.create()
            except HTTPStatusError as error:
                if error.response.status_code == 409:  # noqa: PLR2004
                    logger.warning("Component Conflict, retrying: {}, {}", component.name, component.kind)
                    logger.info(error.response.text)
                    input("Press Enter to continue...")

    Namespace(name=namespace_name).namespace.delete()
```

Running this in a `while True` loop, I'll eventually get a message like:

```
2023-10-12 17:32:16.811 | INFO     | __main__:main:65 - Deploying: eos-migrator, Job
2023-10-12 17:32:16.887 | WARNING  | __main__:main:70 - Component Conflict, retrying: eos-migrator, Job
2023-10-12 17:32:16.888 | INFO     | __main__:main:71 - {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"jobs.batch \"eos-migrator\" already exists","reason":"AlreadyExists","details":{"name":"eos-migrator","group":"batch","kind":"jobs"},"code":409}
```

And if I look in that namespace, I can indeed confirm that the job it's complaining about does indeed exist. So it's almost like the library performed some operation out of order and is now raising a failure. 

I'm not sure what additional information I can provide to help track down the root cause of this, but it seems just running a large number of deployments in a for loop (synchronously) is enough to trigger this eventually.

**Comment by jacobtomlinson** [comment]
Hey there, thanks for reporting this. I'm afraid I find your bug report a little hard to follow, you show examples of a library called `starbug` that I've not heard of and show a `Redis` class that I'm not sure is related or not. Could you have a go at creating a [minimal code snippet](https://matthewrocklin.com/minimal-bug-reports) that I can run and reproduce the problem?

It sounds like you're creating a bunch of resources quickly and running into conflicts. I had a quick go at putting together a reproducer by creating 1000 services in a for loop but it doesn't recreate the problem.

```python
import uuid
from kr8s.objects import Service

# Create a bunch of services very quickly
for _ in range(1000):
    Service({"apiVersion": "v1",
             "kind": "Service",
             "metadata": {"name": "test-" + uuid.uuid4().hex[:6]},
             "spec": {"ports": [{"port": 123, "targetPort": 123}], "selector": {"foo": "bar"}},
        }).create()

print(len(kr8s.get("services")))  # 1001 (the services plus the default kubernetes service)

# Delete them again
for service in kr8s.get("services"):
    if "test" in service.name:
        service.delete()
```

Any further info you can provide to help me reproduce the problem will be important to get to the bottom of this.

**Comment by jacobtomlinson** [comment]
Oh I updated my example into an infinite loop and eventually got one (after 10 mins or so of running).

```python
while True:
    service = Service({"apiVersion": "v1",
             "kind": "Service",
             "metadata": {"name": "test-" + uuid.uuid4().hex[:6]},
             "spec": {"ports": [{"port": 123, "targetPort": 123}], "selector": {"foo": "bar"}},
         })
    service.create()
    print(f"Created service {service.name}")
    service.delete()
    print(f"Deleted service {service.name}")
```

```pytb
---------------------------------------------------------------------------
HTTPStatusError                           Traceback (most recent call last)
Cell In[49], line 7
      1 while True:
      2     service = Service({"apiVersion": "v1",
      3          "kind": "Service",
      4          "metadata": {"name": "test-" + uuid.uuid4().hex[:6]},
      5             "spec": {"ports": [{"port": 123, "targetPort": 123}], "selector": {"foo": "bar"}},
      6         })
----> 7     service.create()
      8     print(f"Created service {service.name}")
      9     service.delete()

File ~/Projects/kr8s-org/kr8s/kr8s/_io.py:46, in run_sync.<locals>.wrapped(*args, **kwargs)
     44 with anyio.from_thread.start_blocking_portal() as portal:
     45     if inspect.iscoroutinefunction(coro):
---> 46         return portal.call(wrapped)
     47     if inspect.isasyncgenfunction(coro):
     48         return iter_over_async(wrapped)

File ~/miniconda3/envs/kr8s/lib/python3.10/site-packages/anyio/from_thread.py:277, in BlockingPortal.call(self, func, *args)
    264 def call(
    265     self, func: Callable[..., Awaitable[T_Retval] | T_Retval], *args: object
    266 ) -> T_Retval:
    267     """
    268     Call the given function in the event loop thread.
    269 
   (...)
    275 
    276     """
--> 277     return cast(T_Retval, self.start_task_soon(func, *args).result())

File ~/miniconda3/envs/kr8s/lib/python3.10/concurrent/futures/_base.py:458, in Future.result(self, timeout)
    456     raise CancelledError()
    457 elif self._state == FINISHED:
--> 458     return self.__get_result()
    459 else:
    460     raise TimeoutError()

File ~/miniconda3/envs/kr8s/lib/python3.10/concurrent/futures/_base.py:403, in Future.__get_result(self)
    401 if self._exception:
    402     try:
--> 403         raise self._exception
    404     finally:
    405         # Break a reference cycle with the exception in self._exception
    406         self = None

File ~/miniconda3/envs/kr8s/lib/python3.10/site-packages/anyio/from_thread.py:217, in BlockingPortal._call_func(self, func, args, kwargs, future)
    214             else:
    215                 future.add_done_callback(callback)
--> 217             retval = await retval
    218 except self._cancelled_exc_class:
    219     future.cancel()

File ~/Projects/kr8s-org/kr8s/kr8s/_objects.py:232, in APIObject.create(self)
    230 async def create(self) -> None:
    231     """Create this object in Kubernetes."""
--> 232     async with self.api.call_api(
    233         "POST",
    234         version=self.version,
    235         url=self.endpoint,
    236         namespace=self.namespace,
    237         data=json.dumps(self.raw),
    238     ) as resp:
    239         self.raw = resp.json()

File ~/miniconda3/envs/kr8s/lib/python3.10/contextlib.py:199, in _AsyncGeneratorContextManager.__aenter__(self)
    197 del self.args, self.kwds, self.func
    198 try:
--> 199     return await anext(self.gen)
    200 except StopAsyncIteration:
    201     raise RuntimeError("generator didn't yield") from None

File ~/Projects/kr8s-org/kr8s/kr8s/_api.py:134, in Api.call_api(self, method, version, base, namespace, url, raise_for_status, stream, **kwargs)
    132         response = await self._session.request(**kwargs)
    133         if raise_for_status:
--> 134             response.raise_for_status()
    135         yield response
    136 except httpx.HTTPStatusError as e:
    137     # If we get a 401 or 403 our credentials may have expired so we
    138     # reauthenticate and try again a few times before giving up.

File ~/miniconda3/envs/kr8s/lib/python3.10/site-packages/httpx/_models.py:749, in Response.raise_for_status(self)
    747 error_type = error_types.get(status_class, "Invalid status code")
    748 message = message.format(self, error_type=error_type)
--> 749 raise HTTPStatusError(message, request=request, response=self)

HTTPStatusError: Client error '409 Conflict' for url 'https://127.0.0.1:37237/api/v1/namespaces/default/services'
For more information check: https://httpstatuses.com/409
```

**Comment by jacobtomlinson** [comment]
No worries @cpressland! It's just really helpful for maintainers to strip out all information and code not related to the bug.

I should have enough info to debug this thanks.

**Comment by jacobtomlinson** [comment]
Quick update. I've been debugging this further and found that this doesn't happen when using the asyncio API, only with the sync one.

It looks like very occasionally it successfully creates the resource but then raises a `RuntimeError` that `The event loop is closed` which causes the `call_api` method to retry. Then we see the 409 on the next try.

https://github.com/kr8s-org/kr8s/blob/4223da89c091cae42112f07586a0ba46732f05f6/kr8s/_api.py#L156-L167

Next we need to establish why the exception is being raised and whether there is some way we can tell if the request succeeded anyway.

**Comment by cpressland** [comment]
Thanks @jacobtomlinson - I should be able to replicate this pretty quick if it does indeed still exist. Thank you so much.
